pytorch_lightning/core/decorators.py

```diff
@@ -64,3 +65,38 @@ def auto_transfer_args(self, *args, **kwargs):
         return fn(self, *args, **kwargs)
 
     return auto_transfer_args
+
+
+def parameter_validation(fn: Callable) -> Callable:
+    """
+    Decorator for :meth:`~pytorch_lightning.core.LightningModule.to` method.
+    Validates that the module parameter lengths match after moving to the device. It is useful
+    when tying weights on TPU's.
+
+    Args:
+        fn: ``.to`` method
+
+    Note:
+        TPU's require weights to be tied/shared after moving the module to the device.
+        Failure to do this results in the initialization of new weights which are not tied.
+        To overcome this issue, weights should be tied using the ``on_post_move_to_device`` model hook
+        which is called after the module has been moved to the device.
+
+    See Also:
+        - `XLA Documentation <https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#xla-tensor-quirks>`_
+    """
+    @wraps(fn)
+    def inner_fn(self, *args, **kwargs):
+        pre_param_count = len(list(self.parameters()))
```

Comment:
shall we use total params instead of total layers?
something like:
```python
pre_param_count = sum(p.numel() for p in self..parameters())
```

Reply:
I think layers themselves should suffice. Is there are use case where we should check the actual total parameter count itself?

Reply:
haven't seen any.. it''s just that total parameter count sounds more like the actual sum of parameters present in the model, not just the layers.

Reply:
I'll change the variable names then. Calculating actual parameter count will be more costly.